src/core/server/integration_tests/saved_objects/migrations/wait_for_migration_completion.test.ts

```diff
@@ -0,0 +1,156 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import Path from 'path';
+import fs from 'fs/promises';
+import JSON5 from 'json5';
+import { kibanaPackageJson as pkg } from '@kbn/utils';
+import { retryAsync } from '@kbn/core-saved-objects-migration-server-mocks';
+import * as kbnTestServer from '../../../../test_helpers/kbn_server';
+import { Root } from '../../../root';
+
+const logFilePath = Path.join(__dirname, 'wait_for_migration_completion.log');
+
+async function removeLogFile() {
+  // ignore errors if it doesn't exist
+  await fs.unlink(logFilePath).catch(() => void 0);
+}
+
+describe('migration with waitForCompletion=true', () => {
+  let esServer: kbnTestServer.TestElasticsearchUtils;
+  let root: Root;
+
+  beforeAll(async () => {
+    await removeLogFile();
+  });
+
+  afterAll(async () => {
+    if (root) {
+      await root.shutdown();
+    }
+    if (esServer) {
+      await esServer.stop();
+    }
+
+    await new Promise((resolve) => setTimeout(resolve, 10000));
+  });
+
+  it('waits for another instance to complete the migration', async () => {
+    const { startES } = kbnTestServer.createTestServers({
+      adjustTimeout: (t: number) => jest.setTimeout(t),
+      settings: {
+        es: {
+          license: 'basic',
+        },
+      },
+    });
+
+    root = createRoot();
+
+    esServer = await startES();
+    await root.preboot();
+    await root.setup();
+
+    root.start();
+    const esClient = esServer.es.getClient();
+
+    await retryAsync(
+      async () => {
+        const logFileContent = await fs.readFile(logFilePath, 'utf-8');
+        const records = logFileContent
+          .split('\n')
+          .filter(Boolean)
+          .map((str) => JSON5.parse(str)) as any[];
+
+        expect(
+          records.find((rec) =>
+            rec.message.startsWith(
+              `[.kibana] Migration required. Waiting until another Kibana instance completes the migration.`
+            )
+          )
+        ).toBeDefined();
+
+        expect(
+          records.find((rec) =>
+            rec.message.startsWith(`[.kibana] INIT -> WAIT_FOR_MIGRATION_COMPLETION`)
+          )
+        ).toBeDefined();
+
+        expect(
+          records.find((rec) =>
+            rec.message.startsWith(
+              `[.kibana] WAIT_FOR_MIGRATION_COMPLETION -> WAIT_FOR_MIGRATION_COMPLETION`
+            )
+          )
+        ).toBeDefined();
+      },
+      { retryAttempts: 100, retryDelayMs: 200 }
+    );
+
+    const aliases: Record<string, {}> = { '.kibana': {} };
+    aliases[`.kibana_${pkg.version}`] = {};
+    await esClient.indices.create({ index: `.kibana_${pkg.version}_001`, aliases });
```

Comment:
NIT: could have simulated a more 'real' environment by booting up 2 `Root` instances, but your approach is at least simpler to control.